sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/ComponentDescriptor.scala

```diff
@@ -0,0 +1,340 @@
+  // once we have built the full file descriptor, we can look up for the input message using its name
+  private case class NamedComponentMethod(
+      serviceMethod: ServiceMethod,
+      grpcMethodName: String,
+      extractorCreators: Map[Int, ExtractorCreator],
+      inputMessageName: String,
+      inputProto: Option[DescriptorProto]) {
+
+    type ParameterExtractors = Array[ParameterExtractor[InvocationContext, AnyRef]]
+
+    def toComponentMethod(fileDescriptor: FileDescriptor): ComponentMethod = {
+      serviceMethod match {
+        case method: SpringRestServiceMethod =>
+          val message = fileDescriptor.findMessageTypeByName(inputMessageName)
+          if (message == null)
+            throw new RuntimeException(
+              "Unknown message type [" + inputMessageName + "], known are [" + fileDescriptor.getMessageTypes.asScala
+                .map(_.getName) + "]")
+
+          val parameterExtractors: ParameterExtractors =
+            if (method.callable) {
+              method.params.zipWithIndex.map { case (param, idx) =>
+                extractorCreators.find(_._1 == idx) match {
+                  case Some((_, creator)) =>
+                    creator(message)
+                  case None =>
+                    // Yet to resolve this parameter, resolve now
+                    param match {
+                      case hp: HeaderParameter =>
+                        new HeaderExtractor[AnyRef](hp.name, identity)
+                      case UnhandledParameter(p) =>
+                        throw new RuntimeException(
+                          s"Unhandled parameter for [${serviceMethod.methodName}]: [$p], message type: " + inputMessageName)
+                      // FIXME not handled: BodyParameter(_, _), PathParameter(_, _), QueryParamParameter(_, _)
+                    }
+                }
+              }.toArray
+            } else Array.empty
+
+          ComponentMethod(serviceMethod.javaMethodOpt, grpcMethodName, parameterExtractors, message)
+
+        case method: AnyServiceMethod =>
+          // methods that receive Any as input always default to AnyBodyExtractor
+          val parameterExtractors: ParameterExtractors = Array(
+            new ParameterExtractors.AnyBodyExtractor(method.inputType))
+          ComponentMethod(serviceMethod.javaMethodOpt, grpcMethodName, parameterExtractors, JavaPbAny.getDescriptor)
+      }
+
+    }
+  }
+
+  private def buildSyntheticMessageAndExtractors(
+      nameGenerator: NameGenerator,
+      serviceMethod: SpringRestServiceMethod,
+      httpRule: HttpRule.Builder,
+      entityKeys: Seq[String] = Seq.empty): (DescriptorProto, Map[Int, ExtractorCreator]) = {
+
+    // FIXME the name depends on order of calls to name generator which could be problematic
+    // (for example sharding passing the request to a different Kalix node where the name was generated
+    // to have 2 instead of 1 in the msg name because of class path ordering not being consistent?)
```

Comment:
Good point. 
I wasn't happy with the NameGenerator using a global counter. It feels weird and unsafe. 
Shall append a hash build from method name and all it's params names? Maybe worth to check how the `Method` hashCode is build.

Reply:
hmm, not unique enough: 

```java
/**
     * Returns a hashcode for this {@code Method}.  The hashcode is computed
     * as the exclusive-or of the hashcodes for the underlying
     * method's declaring class name and the method's name.
     */
    public int hashCode() {
        return getDeclaringClass().getName().hashCode() ^ getName().hashCode();
    }
```

Overloaded methods will have the same hashCode

Reply:
Ok, wont methods named the same in different components end up with the same request message name and clash then?

Reply:
They will be isolated by the service descriptor. Each java class has its own generated descriptor containing all the messages. In that case, I think we are fine.